### ISSUE TITLE
Unhandled Exception when calling submit

### ISSUE THREAD

**Issue by peterstojanowski** [GiancarloCode/form_bloc#63]
Hello,

I just upgraded flutter_form_bloc from version 0.10.0 to 0.12.2 and getting exceptions when calling submit. I noticed that this only happens when I have 3 or more `TextFieldBlocBuilder`s on screen and updating their initial values. Then when I call submit, the app hangs for a few seconds and I see exceptions in the debug console:

```
E/flutter (15380): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: Stack Overflow
E/flutter (15380): #0      new DiagnosticsProperty 
package:flutter/…/foundation/diagnostics.dart:2545
E/flutter (15380): #1      new _ErrorDiagnostic 
package:flutter/…/foundation/assertions.dart:31
E/flutter (15380): #2      new ErrorDescription 
package:flutter/…/foundation/assertions.dart:111
E/flutter (15380): #3      ChangeNotifier.notifyListeners 
package:flutter/…/foundation/change_notifier.dart:212
E/flutter (15380): #4      TextEditingController.selection= 
package:flutter/…/widgets/editable_text.dart:275
E/flutter (15380): #5      _TextFieldBlocBuilderState._fixControllerTextValue 
package:flutter_form_bloc/src/text_field_bloc_builder.dart:730
E/flutter (15380): #6      _TextFieldBlocBuilderState._fixControllerTextValue (package:flutter_form_bloc/src/text_field_bloc_builder.dart)
E/flutter (15380): #7      _TextFieldBlocBuilderState._textControllerListener 
package:flutter_form_bloc/src/text_field_bloc_builder.dart:713
E/flutter (15380): #8      ChangeNotifier.notifyListeners 
package:flutter/…/foundation/change_notifier.dart:206
E/flutter (15380): #9      _TextFieldBlocBuilderState._fixControllerTextValue 
package:flutter_form_bloc/src/text_field_bloc_builder.dart:267
E/flutter (15380): #10     _TextFieldBlocBuilderState._fixControllerTextValue (package:flutter_form_bloc/src/text_field_bloc_builder.dart)
E/flutter (15380): #11     _TextFieldBlocBuilderState._textControllerListener 
package:flutter_form_bloc/src/text_field_bloc_builder.dart:713
E/flutter (15380): #12     ChangeNotifier.notifyListeners 
package:flutter/…/foundation/change_notifier.dart:206
E/flutter (15380): #13     TextEditingController.selection= 
package:flutter/…/widgets/editable_text.dart:275
E/flutter (15380): #14     _TextFieldBlocBuilderState._fixControllerTextValue 
package:flutter_form_bloc/src/text_field_bloc_builder.dart:730
E/flutter (15380): #15     _TextFieldBlocBuilderState._fixControllerTextValue (package:flutter_form_bloc/src/text_field_bloc_builder.dart)
```

Here's my Bloc:
```
class ClientFormBloc extends FormBloc<String, String> {
  final ClientsRepository clientsRepository;
  final Client client;

  final name = TextFieldBloc(
    validators: [CustomValidator.name],
  );
  final email = TextFieldBloc(
    validators: [CustomValidator.emailOptional],
  );
  final phone = TextFieldBloc();
  final website = TextFieldBloc();
  final notes = TextFieldBloc();

  ClientFormBloc({@required this.clientsRepository, @required this.client})
      : super(autoValidate: false) {
    if (client != null) {
      name.updateInitialValue(client.name);
      email.updateInitialValue(client.email);
      phone.updateInitialValue(client.phone);
      website.updateInitialValue(client.website);
      notes.updateInitialValue(client.notes);
    }

    addFieldBlocs(fieldBlocs: [
      name,
      email,
      country,
      phone,
      website,
      notes,
    ]);
  }
  
  @override
  void onSubmitting() async {
    final clientUid = client != null ? client.uid : Uuid().v4();
    final nameValue = name.value.parse();
    final emailValue = email.value.parse();
    final phoneValue = phone.value.parse();
    final websiteValue = website.value.parse();
    final notesValue = notes.value.parse();

    final updatedClient = Client(
        uid: clientUid,
        name: nameValue,
        email: emailValue,
        phone: phoneValue,
        website: websiteValue,
        notes: notesValue);
    clientsRepository.saveClient(updatedClient);
    emitSuccess();
  }
}
```

And the screen widget:
```
class ClientFormScreen extends StatefulWidget {
  ClientFormScreen({key}) : super(key: key);

  @override
  _ClientFormScreenState createState() => _ClientFormScreenState();
}

class _ClientFormScreenState extends State<ClientFormScreen> {
  @override
  Widget build(BuildContext context) {
    final Client client = ModalRoute.of(context).settings.arguments;

    return BlocProvider<ClientFormBloc>(
      create: (context) => ClientFormBloc(
        clientsRepository: RepositoryProvider.of<ClientsRepository>(context),
        client: client,
      ),
      child: Builder(
        builder: (context) {
          final formBloc = context.bloc<ClientFormBloc>();

          return Scaffold(
            appBar: AppBar(
              leading: IconButton(
                icon: Icon(
                  Icons.close,
                ),
                onPressed: Navigator.of(context).pop,
              ),
              title: Text(
                client == null
                    ? tr('general.new_client')
                    : tr('general.edit_client'),
              ),
              actions: <Widget>[
                FlatButton(
                  onPressed: context.bloc<ClientFormBloc>().submit,
                  child: Text(
                    tr('general.save'),
                    style: TextStyle(
                        color: Theme.of(context).colorScheme.onPrimary),
                  ),
                ),
              ],
            ),
            body: FormBlocListener<ClientFormBloc, String, String>(
              onSuccess: (context, state) {
                Navigator.of(context).pop();
              },
              child: ListView(
                children: <Widget>[
                  TextFieldBlocBuilder(
                    textFieldBloc: formBloc.name,
                    keyboardType: TextInputType.text,
                    decoration: InputDecoration(
                      border: OutlineInputBorder(),
                      icon: Icon(MaterialIcons.person_outline),
                      labelText: tr('general.name'),
                    ),
                  ),
                  TextFieldBlocBuilder(
                    textFieldBloc: formBloc.email,
                    keyboardType: TextInputType.emailAddress,
                    decoration: InputDecoration(
                      border: OutlineInputBorder(),
                      icon: Icon(MaterialIcons.mail_outline),
                      labelText: tr('general.email'),
                    ),
                  ),
                  TextFieldBlocBuilder(
                    textFieldBloc: formBloc.phone,
                    keyboardType: TextInputType.phone,
                    decoration: InputDecoration(
                      border: OutlineInputBorder(),
                      icon: Icon(MaterialIcons.phone),
                      labelText: tr('general.phone'),
                    ),
                  ),
                  TextFieldBlocBuilder(
                    textFieldBloc: formBloc.website,
                    keyboardType: TextInputType.url,
                    decoration: InputDecoration(
                      border: OutlineInputBorder(),
                      icon: Icon(MaterialIcons.public),
                      labelText: tr('general.website'),
                    ),
                  ),
                  TextFieldBlocBuilder(
                    textFieldBloc: formBloc.notes,
                    keyboardType: TextInputType.multiline,
                    decoration: InputDecoration(
                      border: OutlineInputBorder(),
                      icon: Icon(MaterialIcons.subject),
                      labelText: tr('general.notes'),
                    ),
                  ),
                ],
              ),
            ),
          );
        },
      ),
    );
  }
}
```

Is there anything I'm doing wrong? I didn't have any issues with the previous version.

Thanks

**Comment by GiancarloCode** [comment]
Hello, a first glance everything looks good.
And I tried as you said with 3 or more field blocs and updating its initial value, but it works for me without any error.
If you can create a small example that has the error, or give me access to your project I can gladly review :).

Also the `country` field of your form bloc does not exist :)

**Comment by vasilich6107** [comment]
Hi there! 👋 
This could be easily reproduced in 'Loading and Initializing' sample just add this

```
//      text.updateInitialValue('I am prefilled');
      text.updateInitialValue(null);
```

and you will get what you are looking for) https://tppr.me/eG9Z2

I assume that this is a regular case when you receive a `null` value from backend)

**Comment by peterstojanowski** [comment]
Yep, that was the case in project as well. Passing `null` to `updateInitialValue` throws an exception.

**Comment by GiancarloCode** [comment]
Thanks for finding the bug @vasilich6107 

I can solve it in 2 ways:
1. When the `textFieldBloc` will receive a null value it will return it to empty string.
2 The `textFierldBlocBuilder` assumes null values as an empty String.

I think the first option is better, I don't think a text field should have null values, it also allows that in custom widgets you do not have to do checks for null values, but I would like to know your options @peterstojanowski @vasilich6107

**Comment by vasilich6107** [comment]
I need to check how default textfield behaves with null values...

**Comment by vasilich6107** [comment]
@GiancarloCode 

Could you clarify what are the reasons behind disabling editing on submitting?

Technically I understand your motivation. Could it be better to let the library user decide to disable the editing on submitting?

I'm asking this cause if we remove this listener the error fixes automatically and we give a little bit more control to the programmer rather than trying to predict their needs.
From the other perspective the implemented solution of disabling by overriding is a little bit unexpected :-)

```
/// Disable editing when the state of the FormBloc is [FormBlocSubmitting].
  void _textControllerListener() {
    if (widget.textFieldBloc.state.formBloc?.state is FormBlocSubmitting) {
      if (_controller.text.isNotEmpty && _controller.text != widget.textFieldBloc.value) {
        _fixControllerTextValue(widget.textFieldBloc.value);
      }
    }
  }
```

**Comment by GiancarloCode** [comment]
@vasilich6107 
At the moment it does that since the field bloc does not allow changing the value if the state is Submitting, so it is for sync the controller with the field bloc.

And in the field bloc I disable update value because I wanted to avoid that when you called onSubmitting you would get a different value, (for example an invalid value), in case you took time to get the values and change the value.
```dart
void onSubmitting() async {
  await someAsyncMethod():
  // while the value change to invalid value;
 field.value; // invalid value :(
}
```
Although I doubt that this is a common case, 
And from 0.11.0 the form bloc state has an immutable state with each value of each field in this moment, although it is necessary to use the 'name' to access it.
```dart
void onSubmitting() async {
  await someAsyncMethod():
  // while the value change to invalid value;
 state.valueOf('fieldName'); // valid value :)
}
```
I think I can allow to change the value when the state is submitting

And add a property to the widget so that it doesn't allow changing the value if the state is submitting.
```dart
TextFieldBloc(canEditWhenIsSubmitting : false) // default true.
```
___
P.D. The error was an infinite loop, solved with this:
```dart
...
   if (_controller.text != (widget.textFieldBloc.value ?? '')) {
...
```
___


Do you have any comments?

**Comment by vasilich6107** [comment]
Technically I'm not a fan of `canEditWhenIsSubmitting` at all.

Here is my why:
- this is a feature a little bit aside either from the form engine or field wrapper
- it will require some efforts on implementation, maintenance and bugfixing
- it is not obvious to user from UI perspective(won't be widely used). When field greyed out or overlayed  during submit seems more user friendly.

Suggested solution with `canEditWhenIsSubmitting` which is `true` by default should not bother anyone until someone decides to use it so it is good to go :-)

**Comment by GiancarloCode** [comment]
Solved in `flutter_form_bloc: 0.13.0`.

* `canEditWhenIsSubmitting` at the moment it is not added
